### PR TITLE
Add prefetch functionality for paginator

### PR DESCRIPTION
**Pull request by zychen5186**
# TL;DR
The get function takes more time to fetch data from the remote server as the amount increases, hence we only fetch 100 rows initially, and prefetch 100 rows of data at once while the user scrolls through pages to enhance the user experience. 

## Type
- [ ] Bug Fix
- [X] Feature
- [ ] Plugin

## Are all requirements met?

- [X] Code completed
- [X] Smoke tested
- [ ] Unit tests added
- [ ] Code documentation added
- [ ] Any pending items have an associated Issue

## Complete description
For testing run:
```bash
make compile
./bin/flytectl get execution -p my-project -d development -i
```

We initially fetch the first 100 rows when the paginator is called and start to prefetch the next batch (next 100 rows).
Prefetch will be triggered when the current page exceeds a prefetch threshold.
The first or last batch will be discarded when it exceeds the cache limit (currently set to 1000 rows).
We added a loading animation to indicate the fetching process is ongoing.
Apr-27-2024 17-32-22

Users can use --filter.page flag to specify their starting page, if not specified, page will start from 1.
Apr-28-2024 16-21-42

## Tracking Issue
https://github.com/flyteorg/flyte/issues/4440

## Follow-up issue
Follow-up of https://github.com/flyteorg/flytectl/pull/473

Big thanks to @troychiu for discussing and debugging with me!! :raised_hands: :heart:

### CODE REVIEW

File: flytectl/pkg/bubbletea/bubbletea_pagination_util.go [hunk 1]
```diff
@@ -84,53 +92,59 @@ func getTable(m *pageModel) (string, error) {
...
+const (
+	forward direction = iota
+	backward
+)
+
+type newDataMsg struct {
+	newItems       []proto.Message
+	batchIndex     int
+	fetchDirection direction
 }
```
Comment (troychiu) [inline flytectl/pkg/bubbletea/bubbletea_pagination_util.go:126]: move up

File: flytectl/pkg/bubbletea/bubbletea_pagination_util.go [hunk 2]
```diff
@@ -84,53 +92,59 @@ func getTable(m *pageModel) (string, error) {
...
+}
+
+func getLocalLastPage() int {
+	sum := 0
+	for i := 0; i < lastBatchIndex+1; i++ {
+		length, ok := batchLen[i]
+		if ok {
+			sum += length
+		} else {
+			sum += msgPerBatch
```
[the remark this answers is not in the document]
Comment (zychen5186) [reply flytectl/pkg/bubbletea/bubbletea_pagination_util.go:146]: I changed a better function name, sorry for misleading :sweat:

**Review by pingsutw — DISMISSED** [review]

**Comment by pingsutw** [comment]
@zychen5186 One of the flytectl tests is failing. Mind taking a look?

**Comment by zychen5186** [comment]
> @zychen5186 One of the flytectl tests is failing. Mind taking a look?

No problem

**Review by troychiu — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
